CRM/Upgrade/Incremental/php/FiveTwelve.php

```diff
@@ -84,4 +96,66 @@ public function setPostUpgradeMessage(&$postUpgradeMessage, $rev) {
   //   return TRUE;
   // }
 
+  /**
+   * Upgrade function.
+   *
+   * @param string $rev
+   */
+  public function upgrade_5_12_alpha1($rev) {
+    $this->addTask(ts('Upgrade DB to %1: SQL', array(1 => $rev)), 'runSql', $rev);
+    $this->addTask('Migrate to utf8mb4', 'migrateUtf8mb4');
+  }
+
+  /**
+   * Performs the utf8mb4 migration.
+   */
+  public static function migrateUtf8mb4(CRM_Queue_TaskContext $ctx) {
+    $dao = new CRM_Core_DAO();
+    $database = $dao->_database;
+    CRM_Core_DAO::executeQuery("ALTER DATABASE $database CHARACTER SET = utf8mb4 COLLATE = utf8mb4_unicode_ci");
+    $dao = CRM_Core_DAO::executeQuery("SHOW TABLE STATUS WHERE (Engine = 'InnoDB' OR Engine = 'ARCHIVE') AND (Name LIKE 'civicrm\_%' OR Name LIKE 'log\_civicrm\_%')");
```

Comment:
For log tables I think we need to handle the case where installations use a separate DB via `CIVICRM_LOGGING_DSN`

Reply:
👍🏽 I'm not sure how to run queries on the logging DSN, lmk if there's a correct way or I can hack something together.

Reply:
Looks like the common approach in core is to [parse the DSN](https://github.com/civicrm/civicrm-core/blob/8b9a8f4abb25cfa7a498eaaa3e459e4d313e350a/CRM/Logging/Differ.php#L49) and then [use it the database as a prefix for all tables in queries](https://github.com/civicrm/civicrm-core/blob/8b9a8f4abb25cfa7a498eaaa3e459e4d313e350a/CRM/Logging/Differ.php#L373). I guess you can use that with `SHOW TABLE STATUS IN $logdb ...`

Reply:
Should be resolved by 66ca5d74516db7cfc88ebaddeeef6b22821c6433